tutorial/lesson_20_cloning_funcs.cpp

```diff
@@ -0,0 +1,143 @@
+        //     g(x, y) = 2 * f(x, y) + 3
+        // for y:
+        //   for x:
+        //     h(x, y) = f(x, y) + g(x, y) + 10
+
+        // Using Func::clone, we can replace calls to 'f' inside 'g' with
+        // clone of 'f' using the schedule alone:
+        Func f_clone_in_g = f.clone_in(g);
+        f_clone_in_g.compute_root();
+
+        // Equivalently, we could also chain the schedules like so:
+        // f.clone_in(g).compute_root();
+
+        // This produces the following loop nests:
+        // for y:
+        //   for x:
+        //     f(x, y) = x + y
+        // for y:
+        //   for x:
+        //     f_clone_in_g(x, y) = x + y
+        // for y:
+        //   for x:
+        //     g(x, y) = 2 * f_clone_in_g(x, y) + 3
+        // for y:
+        //   for x:
+        //     h(x, y) = f(x, y) + g(x, y) + 10
+
+        h.realize(5, 5);
+
+        // The schedule directive f.clone_in(g) replaces all calls to 'f'
+        // inside 'g' with a clone of 'f' and then returns that clone.
+        // Essentially, it rewrites the original pipeline above into the
+        // following:
+        {
+            Func f_clone_in_g("f_clone_in_g"), f("f"), g("g"), h("h");
+            f(x, y) = x + y;
+            f_clone_in_g(x, y) = x + y;
+            g(x, y) = 2 * f_clone_in_g(x, y) + 3;
+            h(x, y) = f(x, y) + g(x, y) + 10;
+
+            f.compute_root();
+            f_clone_in_g.compute_root();
+            g.compute_root();
+            h.compute_root();
+        }
+    }
+
+    {
+        // In the schedule above, only the calls to 'f' made by 'g' are
+        // replaced. Other calls made to 'f' would still call 'f' directly
+        // (i.e. 'h' still calls 'f' and not the clone). If we wish to
+        // replace all calls to 'f' made by both 'g' and 'h' with a single
+        // clone, we simply say f.clone_in({g, h}).
+
+        // Consider a three stage pipeline, with two consumers of f:
+        Func f("f_group"), g("g_group"), h("h_group"), out("out_group");
+        f(x, y) = x + y;
+        g(x, y) = 2 * f(x, y);
+        h(x, y) = f(x, y) + 10;
+        out(x, y) = f(x, y) + g(x, y) + h(x, y);
+
+        f.compute_root();
+        g.compute_root();
+        h.compute_root();
+        out.compute_root();
+
+        // We will replace all calls to 'f' inside both 'g' and 'h'
+        // with calls to a single clone:
+        f.clone_in({g, h}).compute_root();
+
+        // The equivalent loop nests are:
+        // for y:
+        //   for x:
+        //     f(x, y) = x + y
+        // for y:
+        //   for x:
+        //     f_clone(x, y) = x + y
+        // for y:
+        //   for x:
+        //     g(x, y) = 2 * f_clone(x, y)
+        // for y:
+        //   for x:
+        //     h(x, y) = f_clone(x, y) + 10
+        // for y:
+        //   for x:
+        //     out(x, y) = f(x, y) + g(x, y) + h(x, y)
+
+        out.realize(5, 5);
+    }
+
+    printf("Success!\n");
+
+    return 0;
```

Comment:
The thing that's missing here is an example of a realistic use-case. I guess one might be if two consumers of a Func need regions of it that are disjoint and far apart, then you'd be better of with two different copies of the producer.